src/transports/ssh_libssh2.c

```diff
@@ -0,0 +1,376 @@
+struct git_ssh__sign_payload
+{
+	git_cred_sign_cb sign_cb;
+	void *user_payload;
+};
+
+static inline LIBSSH2_USERAUTH_PUBLICKEY_SIGN_FUNC(git_ssh__sign_cb)
+{
+	struct git_ssh__sign_payload *payload = *abstract;
+
+	return payload->sign_cb((git_ssh_session *)session,
+							sig, sig_len,
+							data, data_len, payload->user_payload);
+}
+
+struct git_ssh__kbdint_payload
+{
+	LIBSSH2_SESSION *session;
+	git_cred_ssh_interactive_cb kbdint_cb;
+	void *user_payload;
+};
+
+static inline LIBSSH2_USERAUTH_KBDINT_RESPONSE_FUNC(git_ssh__kbdint_cb)
+{
+	struct git_ssh__kbdint_payload *payload = *abstract;
+
+	GIT_UNUSED(prompts);
+	GIT_UNUSED(responses);
+
+	return payload->kbdint_cb((git_ssh_session *)payload->session,
+							  name, name_len,
+							  instruction, instruction_len,
+							  num_prompts, NULL, NULL,
+							  payload->user_payload);
+}
+
+int _git_ssh_authenticate_session(git_ssh_session *s, git_cred *cred)
+{
+	int rc;
+
+	do {
+		git_error_clear();
+		switch (cred->credtype) {
+			case GIT_CREDTYPE_USERPASS_PLAINTEXT: {
+				git_cred_userpass_plaintext *c = (git_cred_userpass_plaintext *)cred;
+				rc = libssh2_userauth_password(s->session, c->username, c->password);
+				break;
+			}
+			case GIT_CREDTYPE_SSH_KEY: {
+				git_cred_ssh_key *c = (git_cred_ssh_key *)cred;
+
+				if (c->privatekey) {
+					rc = libssh2_userauth_publickey_fromfile(
+						s->session, c->username, c->publickey,
+						c->privatekey, c->passphrase);
+				} else
+					rc = git__ssh_agent_auth(s, c);
+
+				break;
+			}
+			case GIT_CREDTYPE_SSH_CUSTOM: {
+				git_cred_ssh_custom *c = (git_cred_ssh_custom *)cred;
+				struct git_ssh__sign_payload payload = { c->sign_callback, &c->payload };
+
+				rc = libssh2_userauth_publickey(
+					s->session, c->username, (const unsigned char *)c->publickey,
+					c->publickey_len, git_ssh__sign_cb, (void **)&payload);
+				break;
+			}
+			case GIT_CREDTYPE_SSH_INTERACTIVE: {
+				git_cred_ssh_interactive *c = (git_cred_ssh_interactive *)cred;
+				void **abstract = libssh2_session_abstract(s->session);
+				struct git_ssh__kbdint_payload payload;
+				payload.kbdint_cb = c->prompt_callback;
+				payload.session = s->session;
+				payload.user_payload = c->payload;
+
+				/* ideally, we should be able to set this by calling
+				 * libssh2_session_init_ex() instead of libssh2_session_init().
+				 * libssh2's API is inconsistent here i.e. libssh2_userauth_publickey()
+				 * allows you to pass the `abstract` as part of the call, whereas
+				 * libssh2_userauth_keyboard_interactive() does not!
+				 *
+				 * The only way to set the `abstract` pointer is by calling
+				 * libssh2_session_abstract(), which will replace the existing
+				 * pointer as is done below. This is safe for now (at time of writing),
+				 * but may not be valid in future.
+				 */
+				*abstract = &payload;
+
+				rc = libssh2_userauth_keyboard_interactive(
+					s->session, c->username, git_ssh__kbdint_cb);
+
+				*abstract = old_abstract;
```

Comment:
I see you added `old_abstract` in commit 98c9f6a6, but its declaration got lost in the following commit 749f3440.